### PR TITLE
feat(medusa): Performance improvements of Carts domain

### PR DESCRIPTION
**Pull request by adrien2p**
**What**

I have created a new method on the cart service which is `addLineItems`, allowing a user to add one or multiple items in an optimized way. Also updated the `generate` method from the line item service which now also accept a object data or a collection of data which. Various places have been optimized and cache support has been added to the price selection strategy.

The overall optimization allows to reach another 9000% improvement in the response time as a median (Creating a cart with 6 items):

|   | Min (ms)  | Median (ms)  | Max (ms)  | Median Improvement (%)
|---|:-:|---|---|---|
| Before optimisation  | 1200  | 9999 | 12698  |  N/A
| After optimisation | 63  | 252  | 500  | 39x
| After re optimisation | 56 | 82  | 399  | 121x
| After including addressed feedback | 65 | 202  | 495  | 49x

FIXES CORE-722

### CODE REVIEW

**Comment by adrien2p** [comment]
Before refactoring:
Screenshot 2022-11-22 at 11 53 12


After refactoring:
Screenshot 2022-11-22 at 11 52 21

**Comment by adrien2p** [comment]
Last result with the same env

Screenshot 2022-11-23 at 09 35 26

File: packages/medusa/src/services/cache.ts
Comment (olivermrbl) [inline packages/medusa/src/services/cache.ts:5]: praise: This is great!

thought: This couples us to Redis as caching layer. We could leverage our new Module Resolution API to create a CacheAbstraction (similar to the one @carlos-r-l-rodrigues proposed) as well as a module definition. The module could use the Redis service as a default.

Comment (adrien2p) [reply packages/medusa/src/services/cache.ts:5]: In fact, it abstract redis to a single place instead of having redis reference in the entire code base as it is at the moment. If all the places use this service instead of redis directly it means that first it can become a module/strategy and second that if we want to use something else, we just have to change the implementation of the cache service method and it will be applied on the entire code base.

Comment (olivermrbl) [reply packages/medusa/src/services/cache.ts:5]: Not sure I understand.

My idea was to create an `ICacheService` interface exposing three methods; set, get, invalidate - in the first iteration.

Then create a module `@medusajs/cache-redis` similar to the event bus + inventory management proposals, and use that module as default in the module definition. 

That would allow developers to use any cache service, create a module for it, and replace the default Redis one.

Comment (adrien2p) [reply packages/medusa/src/services/cache.ts:5]: Maybe you would prefer that I remove the cache service (which is an intermediary step toward the module) and instead inject the `redisClient` directly in the price selection strategy which means that we coupled another place to it

Comment (adrien2p) [reply packages/medusa/src/services/cache.ts]: Since this is not part of that pr,  I have just created the appropriate interface that the cache service implements. In another pr we will be able to extract it to a separate module as it was the intention behind what I have done but I wanted to keep it simple in that pr. would you like me to go further

Comment (carlos-r-l-rodrigues) [reply packages/medusa/src/services/cache.ts]: Maybe it would be a good idea to let the cache out of this PR completely. You can keep a few `// TODO:` comments to mark the places where we could cache the values.
Ideally we should implement a more robust solution for caching things and do not add much complexity in the code itself with many if/else to check whether it has cache and set it manually.

Comment (adrien2p) [reply packages/medusa/src/services/cache.ts]: yeah that is true, but I need the cache here otherwise I get an improvement of the perf but not enough. Also, this is a place where it is required. My idea was to not complexity to much the price selection strategy and try to not reference anything related to redis and therefore extract that to a service.

Comment (adrien2p) [reply packages/medusa/src/services/cache.ts]: My suggestion was to create a ticket after that PR to extract the cache system. But we can't not use it in the mean time. So this is why I tried to find an intermediary steps where we remove the notion of redis from the services to only have it in one place (the cache service) and therefore simplify the migration later. But in the mean time it still allow us to use it without adding more reference to redis itself into the services

Comment (adrien2p) [reply packages/medusa/src/services/cache.ts]: wdyt @carlos-r-l-rodrigues

Comment (carlos-r-l-rodrigues) [reply packages/medusa/src/services/cache.ts]: I don't have enough knowledge about all the moving pieces composing the price selection, but at a first glance I find the cache key very granular:
`ps:${variantId}:${context.region_id}:${context.currency_code}:${context.customer_id}:${context.quantity}:${context.include_discount_prices}:${taxRate}`

I'm affraid that we will create a gigantic amount of entries if many users are using the system and that will create other type of problems.

Comment (adrien2p) [reply packages/medusa/src/services/cache.ts]: At the moment, if a number of concurrent users are doing it without the cache it causes a lot of problem as well as this will be called a gigantic amount of time as well. I think it is a trade off and i ve benchmarked it with a load test as well and it is much better

Comment (adrien2p) [reply packages/medusa/src/services/cache.ts]: After a discussion with @carlos-r-l-rodrigues we agreed upon the trade off as a temporary solution until the new approach a cache system is available.

**Comment by adrien2p** [comment]
After re optimisation, specifically the generate method

Screenshot 2022-11-23 at 13 45 54

File: packages/medusa/src/interfaces/price-selection-strategy.ts
Comment (srindom) [reply packages/medusa/src/interfaces/price-selection-strategy.ts:61]: An alternative approach would be to make the `get*Price` functions in the PricingService work with bulk inputs so fetching can be done once instead of having to referch the same data multiple times

File: packages/medusa/src/services/index.ts
Comment (srindom) [inline packages/medusa/src/services/index.ts:38]: question: is this intentional?

File: packages/medusa/src/services/cache.ts
Comment (olivermrbl) [reply packages/medusa/src/services/cache.ts:1]: Agree! We had a long discussion earlier today [here](https://github.com/medusajs/medusa/pull/2648#discussion_r1030229206) on how to approach it.

File: packages/medusa/src/interfaces/price-selection-strategy.ts
Comment (adrien2p) [reply packages/medusa/src/interfaces/price-selection-strategy.ts:61]: Yes the bulk is the other option i am looking at like i ve done for the generate and other methods. i completely agree on just passing the data instead of the entire entity btw. I usually do that but ive forgot to change it after all

File: packages/medusa/src/services/index.ts
Comment (adrien2p) [reply packages/medusa/src/services/index.ts:38]: Woups, i missed it. No it is not intentional i ll fix that

File: packages/medusa/src/types/line-item.ts
Comment (adrien2p) [reply packages/medusa/src/types/line-item.ts:16]: It is pulling a lot of strings out of the scope of that pr and I don't want to force cast as it can lead to errors if someone else comes after me. So I did a part of that one that I have spread along the way which is specifying the cart props that are expected but those props are still entity base. I think it is a trade-off on that pr to not become too huge.

**Comment by adrien2p** [comment]
Screenshot 2022-11-24 at 12 45 59

File: packages/medusa/src/services/line-item.ts
Comment (carlos-r-l-rodrigues) [inline packages/medusa/src/services/line-item.ts:249]:
```suggestion
          if (resolvedContext.unit_price == null) {
```

File: packages/medusa/src/types/line-item.ts
Comment (carlos-r-l-rodrigues) [inline packages/medusa/src/types/line-item.ts:8]: we could add a more descriptive type here. Maybe something like `GenerateLineItemContext`

Comment (adrien2p) [reply packages/medusa/src/types/line-item.ts:8]: I though the name of the method would be enough, but I guess you are right ahah

Comment (adrien2p) [reply packages/medusa/src/types/line-item.ts]: @carlos-r-l-rodrigues on another subject I have created a type `CalculationContextData` but I did not change the name of the props in purpose cause I don't want to integrate breaking changes like that in the scope of that pr. wdyt?

**Review by carlos-r-l-rodrigues — APPROVED** [review]

**Review by srindom — COMMENTED** [review]
Great work - had a few thoughts while going over the changes. I don't mind approving as-is if you think it makes sense to move further changes to another PR.

File: packages/medusa/src/services/inventory.ts
Comment (adrien2p) [reply packages/medusa/src/services/inventory.ts:75]: I agree but the thing is that by doing that and not allowing us to pass the data if we have them (it is typed and the auto-completion help avoids mistake) we are increasingly augmenting the number of queries that are duplicated when it is not necessarily needed. The impact of that is increasing the number of queries stacked to the database, increasing the db charge, increasing the data sent through the network, reducing the throughput, and increasing the response time. For simple usage and complexity of the operations, it is still possible to just pass the id. But in those cases where we already have the data, you don't have to pass the entire entity but the data expected by the method. I know it is a trade-off which is why I always allow passing the data or the id. I still want to warn you about that as it creates a lot of overhead and surcharges the system for some operations that we could avoid and will lead to other issues that we already have.

I let you choose but If you want to do that, no problem, I ll operate the changes and will see later when we get back to the problem and that we will have to do such changes.

File: packages/medusa/src/services/line-item.ts
Comment (adrien2p) [reply packages/medusa/src/services/line-item.ts:309]: Partial variant make all the properties optional whereas here we know the props that are not optional. It also helps to auto-document the API of the method and avoid forgetting to pass some data since Partial will accept anything from a variant and you would have to read the method body to know what is needed as a consumer in case you want to select the required fields from the db and hence improve the db queries. wdyt?

File: packages/medusa/src/types/pricing.ts
Comment (adrien2p) [reply packages/medusa/src/types/pricing.ts:37]: you mean just passing the id and not the product_id? If that the case, it is the same answer as above. Accumulating a judge number of queries on top of each other to get the same data as we already hold will only increase the idle transaction time and response time. It is a trade-off and I can do all the changes you've asked and re do a benchmark. I think we will still be better than what we use to have. But, a lot higher than what we can get but let see

File: packages/medusa/src/services/pricing.ts
Comment (adrien2p) [reply packages/medusa/src/services/pricing.ts:252]: Before doing any other changes on that PR I would like to get your opinion on the above proposal so that I can do everything in one go 👍

**Comment by adrien2p** [comment]
@srindom Just to get an idea of the costs of some trade-offs, that I tried to explain, you can see the updated benchmark in the description to see the diff that we go for now.

File: packages/medusa/src/services/inventory.ts
Comment (adrien2p) [reply packages/medusa/src/services/inventory.ts]: @srindom In the mean time I have address that feedback and updated the benchmark

File: packages/medusa/src/services/pricing.ts
Comment (adrien2p) [reply packages/medusa/src/services/pricing.ts:252]: The old method is not used in the core anymore, just wanted to mark it as deprecated since the new one does the same job as well

File: packages/medusa/src/services/inventory.ts
Comment (srindom) [reply packages/medusa/src/services/inventory.ts]: Completely understand the considerations here. I think it is worth it to stick with a slight performance cost for a better API experience so given the benchmarks you just ran I would argue that we allow the extra few queries, but agree that it is something to keep in mind :)

File: packages/medusa/src/services/line-item.ts
Comment (srindom) [reply packages/medusa/src/services/line-item.ts:309]: Makes sense, my point was more that this is quite verbose but fine to keep as-is :)

File: packages/medusa/src/types/pricing.ts
Comment (srindom) [reply packages/medusa/src/types/pricing.ts:37]: my point was more to fetch the product variant within the method and then use `variant.product_id` instead of adding the responsibility to the caller of making sure that `id` and `product_id` are correctly associated with each other. I realize this will again introduce additional DB queries, but similar to the argument made previously it might be worth it to avoid complexity in the API, wdyt?

Comment (adrien2p) [reply packages/medusa/src/types/pricing.ts:37]: I ll try it out and share the result soon 👍

Comment (adrien2p) [reply packages/medusa/src/types/pricing.ts]: I ve done the changes in a bulk way. The perf are a bit better or a bit worst but still as good 🚀

**Comment by adrien2p** [comment]
@olivermrbl @srindom If you want to re-throw an eye. The integration tests have been fixed

**Review by srindom — COMMENTED** [review]
LGTM w. comments - pending @olivermrbl

File: integration-tests/api/__tests__/store/cart/cart.js
Comment (adrien2p) [reply integration-tests/api/__tests__/store/cart/cart.js:38]: oups :/ right away

File: integration-tests/api/__tests__/store/cart/ff-tax-inclusive-pricing.js
Comment (adrien2p) [reply integration-tests/api/__tests__/store/cart/ff-tax-inclusive-pricing.js:30]: I have just defaulted the value to false in the helper. but it is still possible to set it here. We had some suite to false and others to true when it was not necessary. The auto-completion will still propose verbose as an option and we just have to flip it if necessary. wdyt?

File: integration-tests/api/__tests__/admin/product.js
Comment (olivermrbl) [inline integration-tests/api/__tests__/admin/product.js:44]: suggestion(non-blocking): Why is this added? Think a comment here would be appropriate :)

File: integration-tests/api/__tests__/batch-jobs/product/export.js
Comment (olivermrbl) [inline integration-tests/api/__tests__/batch-jobs/product/export.js:36]: Goes with my other comment on the cache ttl

File: integration-tests/api/__tests__/store/cart/ff-tax-inclusive-pricing.js
Comment (olivermrbl) [reply integration-tests/api/__tests__/store/cart/ff-tax-inclusive-pricing.js:30]: I second this. 

For me, the autocompletion is not helpful, but the `"cmd + F" > verbose` operation allowing me to flip the flag quickly is.

File: packages/medusa/src/services/cache.ts
Comment (olivermrbl) [inline packages/medusa/src/services/cache.ts:1]: todo(non-blocking): Can I get you to create a chore issue on replacing the cache implementation in the tax provider to ensure we don't end up with duplicated code?

Comment (olivermrbl) [reply packages/medusa/src/services/cache.ts:1]: Additionally, I think we should tackle that same issue asap to not let it be forgotten.

File: packages/medusa/src/services/cart.ts
Comment (olivermrbl) [inline packages/medusa/src/services/cart.ts:334]: question: Are we not unnecessarily constraining our selves with this validation? If the region id is provided, we will attempt to fetch the region on the lines following this check.

File: packages/medusa/src/services/cart.ts
Comment (olivermrbl) [inline packages/medusa/src/services/cart.ts:560]: suggestion(non-blocking): Extract to a type

File: packages/medusa/src/services/cart.ts
Comment (olivermrbl) [inline packages/medusa/src/services/cart.ts:795]: suggestion(non-blocking): Minor simplification
```suggestion
        const itemsToUpdateKeys = Object.keys(lineItemsToUpdate)
        // Update all items that needs to be updated
        if (itemsToUpdateKeys?.length) {
          await Promise.all(
            itemsToUpdateKeys.map(async (variantId) => {
```

File: packages/medusa/src/services/cart.ts
Comment (olivermrbl) [inline packages/medusa/src/services/cart.ts:815]: question: Why do we fail silently here? If an error happens, the line items will be in an invalid state unknown to the consumer.

File: packages/medusa/src/services/event-bus.ts
Comment (olivermrbl) [inline packages/medusa/src/services/event-bus.ts:64]: comment: Just a note, I think we should be careful about including such changes in a completely unrelated PR. It will be harder to revert in case we experience issues (although unlikely to happen in this case), but I just wanted to mention it. Flipping the verbosity flag is a somewhat similar change.

File: packages/medusa/src/services/pricing.ts
Comment (olivermrbl) [inline packages/medusa/src/services/pricing.ts:322]: comment: Could this be done as part of initializing the map?

File: packages/medusa/src/services/line-item.ts
Comment (olivermrbl) [inline packages/medusa/src/services/line-item.ts:381]: suggestion(non-blocking):
```suggestion
        const data_ = Array.isArray(data) ? data : [data]
```

File: packages/medusa/src/services/line-item.ts
Comment (olivermrbl) [inline packages/medusa/src/services/line-item.ts:234]: suggestion(non-blocking):
```suggestion
        const resolvedData = (
          Array.isArray(data) ? data : [data]
        ) as GenerateInputData[]
```

File: packages/medusa/src/services/cache.ts
Comment (adrien2p) [reply packages/medusa/src/services/cache.ts:1]: I think that it should be replaced everywhere in the core in fact. 
The purpose of that service is to have a first step of removing the notion of redis in the service directly and be localised in one place for later extraction. At the moment, the tax provider is the only one that does no reference redis directly unlike the other places where the cache is used.
What do you think of a linear issue in two steps:
1. replace the cache usage to use the cache service
2. extract it to a module

File: packages/medusa/src/services/cart.ts
Comment (adrien2p) [reply packages/medusa/src/services/cart.ts:334]: here we are checking if none of the id or region is provided. In that case we cannot fetch it without the id

File: packages/medusa/src/services/event-bus.ts
Comment (olivermrbl) [reply packages/medusa/src/services/event-bus.ts:64]: I think it's fine to keep them in, but I just wanted to add it as a note for us to be aware of it going forward 👍

File: packages/medusa/src/services/cart.ts
Comment (adrien2p) [reply packages/medusa/src/services/cart.ts:334]: You don't have to provide both but only one of them, both are optional but at least one must be provided. as usual, it seems like a little perf improvement but with all the queries accumulated it makes a considerable difference.

File: packages/medusa/src/services/cart.ts
Comment (adrien2p) [reply packages/medusa/src/services/cart.ts:815]: What I could do is checking for that specific error and throw otherwise. wdyt @olivermrbl

Comment (adrien2p) [reply packages/medusa/src/services/cart.ts:815]: I would rather be safe too and check for that specific error 💪

**Review by olivermrbl — APPROVED** [review]
LGTM, feel free to merge without addressing last couple of comments :)

File: packages/medusa/src/services/cart.ts
Comment (adrien2p) [reply packages/medusa/src/services/cart.ts:815]: Also, one question i am asking myself, why do the update method throw when there is no element found? I would have thought that we can try to update something and if there is nothing to update then we do nothing. Like and sql query which does not perform an update if there is no tuple. Wdyt?

Comment (olivermrbl) [reply packages/medusa/src/services/cart.ts:815]: I see your point. Let's keep that in mind for a separate PR to avoid having it block this one 👍

File: packages/medusa/src/services/cart.ts
Comment (olivermrbl) [inline packages/medusa/src/services/cart.ts:670]: thought: This does not seem super scalable. Could we check the error type instead?

Comment (adrien2p) [reply packages/medusa/src/services/cart.ts:670]: yes you are right 👍

Comment (srindom) [reply packages/medusa/src/services/cart.ts]: A thought here - does it make sense to have the update function throw a not found error? It is not something we generally do across other services.

**Review by olivermrbl — APPROVED** [review]
LGTM 💪

**Comment by adrien2p** [comment]
I ve closed and reopened the pr to trigger the pipeline to run again

**Pull request merged**
